bind-2/bind/contrib/updatehosts-1.0.0/man/updatehosts.1
.\" @(#) updatehosts.1 1.3 99/01/04 @(#)
.TH UPDATEHOSTS 1 "99/01/04"
.UC 4
.SH NAME
updatehosts \- update DNS and static host tables
.SH SYNOPSIS
.B updatehosts
[
.B -d dbdirectory
]
[
.B -f configfile
]
[
.B -n
]
[
.B -s
]
[
.B -v
]
.B [ file ... ]
.SH DESCRIPTION
Updates the DNS and static host database zone files.
.PP
.B updatehosts
is a front-end to the
.B gendns
and
.B genstatic
programs which generate DNS zone files the and static host file from a set
of input files organized as a set of flat relational files.
.B updatehosts
is driven off of a configuration file usually found in
.B /usr/local/bin/updatehosts.env
which sets variables such as the location of the input
database files.
An alternate configuration file location can be specified by the -f option.
.PP
The input database files are normally kept in the directory
.B /var/named/db.
The location of this directory can be specified by the -d option or with the
DB_DIR environment variable.
Output from
.B updatehosts
is normally placed in the
.B /var/named
directory.
An alternate output directory location may be specified by setting the
NAMED_DIR variable.
.PP
The following files are maintained in the DB_DIR directory:
.TP
main
Contains information about each host within a set of domains.  Information
such as the domain name of each host and its IP address are maintained in
this file.
.TP
cname
Contains information about aliases for hosts
.TP
mx
Contains information about MX (mail exchanger) records.
.TP
ns
Contains information about NS (nameserver) records.
.TP
soa
Contains information about SOA (start of authority) records.
.TP
global
Contains some global parameters
.TP
secondary
Contains information about secondary zones served by this server
.TP
subnet
Contains information about local subnets which is used to generate an
optional BOOTP or DHCP configuration file.
.PP
Each of the database files is organized tabularly.  These files are
preprocessed by the
.B readinfo
command.
(see readinfo.1)
.PP
.B updatehosts
may takes arguments which are the names of the database files which the user
wants to edit.  All database files may optionally kept under SCCS control
by setting the USE_SCCS variable or by placing the -s flag on
the command line.
Similarly RCS control can be maintained by setting the USE_RCS
variable or setting the -r flag on the command line.
A user using
.B updatehosts
must have write permission in the DB_DIR and the NAMED_DIR directory.
Each database file is edited by the
.B vi
editor or whatever program the user has set in the EDITOR environment
variable.
.PP
After each edit, the database files are optionally checked back in using SCCS
or RCS.
Then the static host file is generated using the
.B genstatic
command.
The host file is normally placed in the NAMED_DIR directory under the name
.B static.
This name may be changed by setting the STATICHOSTS variable.
.PP
Then the DNS zone files and
bootstrap file are updated using the
.B gendns
command.
The VERSION variable can be set to either 4 or 8 depending on
the version of the nameserver in use.
This option alters the format of the nameserver bootstrap file.
.PP
The local nameserver is restarted
with the newly updated zone and bootstrap files by calling
.B "poke_ns restart"
.PP
Optionally, the POSTPROCESS environment variable can be set to the name of
a command to run after the nameserver zone files and static host files are
generated.
This can be any arbitrary command.
Normally, this may be a command to update NIS or NIS+ host maps from the
generated static host file.
.PP
Environment variables should normally be set in the configuration file
except when testing.  An exception to this rule is the EDITOR environment
variable which is normally set to the user's favorite editor.
.SH
Normal Configuration Setup
.PP
.B updatehosts
provides a great deal of flexibility in how a nameserver zone files and
static host table are generated and the various permissions on those files.
In normal practice it is desirable to set up a group (say, staff) to manage
the nameserver files.
Setting the directory permissions on the NAMED_DIR tree to allow staff to
create and modify files in that tree allows
.B updatehosts
to be run by anyone in the staff group without root access.
The
.B poke_ns
program must be run as root to restart the nameserver.
So setting it setuid to root, but only executable by the staff group allows
any member of the staff group to restart the running nameserver. 
.SH
See also
.PP
.BR gendns(1),
.BR genstatic(1),
.BR poke_ns(1)


bind-2/bind/contrib/updatehosts-1.0.0/updatehosts.sh
#!/bin/sh
# @(#) updatehosts.sh 1.53 99/01/05 @(#)

# updatehosts DNS maintenance package
# Copyright (C) 1998  Smoot Carl-Mitchell
# 
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
# 
# smoot@tic.com

# update the host tables and DNS files
#
# arguments are the files to edit

CONFIG_FILE=@bindir@/updatehosts.env
DB_DIR=/var/named/db
EDITOR=${EDITOR-vi}
JUST_D=
NAMED_DIR=/var/named
POSTPROCESS=
STATICHOSTS=static
USERCS=
USESCCS=1
VERSION=8

update=1
verbose=
while getopts D:d:f:nsv option; do
	case $option in
	D) $JUST_D=$OPTARG ;;
	d) DB_DIR=$OPTARG ;;
	f) CONFIG_FILE=$OPTARG ;;
	n) update= ;;
	r) USERCS=1 ;;
	s) USESCCS=1 ;;
	v) verbose=-v ;;
	esac
done
shift `expr $OPTIND - 1`

# look for the startup file which sets global environment variables
# this file must be present
if [ ! -f $CONFIG_FILE ]; then
	echo $CONFIG_FILE not found
	exit 1
fi
. $CONFIG_FILE

list=
# check the filename arguments
for i do
	found=
	for j in $DB_FILES; do
		if [ $i = $j ]; then
			found=1
			break
		fi
	done
	if [ ! "$found" ]; then
		echo $i is not a valid filename
		list=1
	fi
done
if [ "$list" ]; then
	echo legal filenames are:
	echo $DB_FILES
	exit 1
fi

cd $NAMED_DIR || ( echo no $NAMED_DIR directory; exit 1 )

# set umask to group read and write
umask 2
# check if the files exist and are under SCCS control and are readable
#
# files are in a subdirectory - so we cd there
cd $DB_DIR || ( echo no $DB_DIR present; exit 1 )

okay=1
if [ "$USESCCS" ]; then for i do
	if [ ! -f SCCS/s.$i ]; then
		echo "file $i is not under SCCS control"
		okay=0
	elif [ ! -r SCCS/s.$i ]; then
		echo "file $i is not readable by you"
		okay=0
	elif [ -f SCCS/p.$i ]; then
		echo "file $i is already checked out under SCCS"
		okay=0
	elif [ -f $i ]; then
		rm -f $i
	fi
done
elif [ "$USERCS" ]; then for i do
	if [ ! -f RCS/$i,v ]; then
		echo "file $i is not under RCS control"
		okay=0
	elif [ ! -r RCS/$i,v ]; then
		echo "file $i is not readable by you"
		okay=0
	elif [ -f $i ]; then
		rm -f $i
	fi
done fi

if [ $okay -eq 0 ]; then
	exit 1
fi

# checkout and edit the files
for i do
	if [ "$USESCCS" ]; then
		sccs edit $i
	elif [ "$USERCS" ]; then
		co -l $i
	fi
	$EDITOR $i
	if [ "$USESCCS" ]; then
		sccs delget $i
	elif [ "$USERCS" ]; then
		ci $i
	fi
done

cd $NAMED_DIR

# generate the static tables
if [ $STATICHOSTS ]; then
	echo "generating static hosts table..."
	HOSTS=$STATICHOSTS genstatic
fi

# generate the dns zone files
echo "generating the DNS database..."
if [ "$JUST_D" ]; then
	gendns $verbose -D "$JUST_D" -d $DB_DIR
else
	gendns $verbose -d $DB_DIR
fi

# generate the bootptab file 
# but only if we are instructed to
if [ $GENBOOTP ]; then
	echo "generating the bootptab file..."
	genbootp

	# copy it to alternate machines
	for m in $SECBOOTP; do
		cp -p bootptab /net/$SECBOOTP/etc
	done
fi

# generate the dhcptab file 
# but only if we are instructed to
if [ $GENDHCP ]; then
	echo "generating the dhcpd.conf file..."
	if [ $verbose ]; then
		verbose=-v
	fi
	if [ $KNOWNHOSTS ]; then
		known=-k
	fi
	gendhcp $verbose $known

	# copy it to alternate machines
	for m in $SECDHCP; do
		cp -p dhcpd.conf /net/$SECDHCP/etc/dhcpd.conf
	done

	# restart the server
	poke_dhcp >/dev/null 2>&1
fi

# run postprocessor command
if [ "$POSTPROCESS" ]; then
	echo running $POSTPROCESS
	eval $POSTPROCESS
fi

if [ $update ]; then
	poke_ns restart
fi
